### PR TITLE
json-bigint based implementation

### LINKED ISSUE #12: Throw JsonDecode errors such that they make sense when being printed
When using `printJsonDecodeError` from argonaut, it adds text to describe the error:
https://github.com/purescript-contrib/purescript-argonaut-codecs/blob/176a5dddae28fda9a2ff31ed4bf1efcc148513a4/src/Data/Argonaut/Decode/Error.purs#L33-L43

Ideally aeson should be throwing errors that are readable when shown via this function, eg `TypeMismatch` should just contain the type name, but here there is a sentence
https://github.com/mlabs-haskell/purescript-aeson/blob/211a4d4e7f48aca48f52bf461d83c2d6f3eec24e/src/Aeson.purs#L491

And here
https://github.com/mlabs-haskell/purescript-aeson/blob/211a4d4e7f48aca48f52bf461d83c2d6f3eec24e/src/Aeson.purs#L520

### PR DESCRIPTION
**Pull request by kirill-havryliuk**
Closes #12, #11 

Full rework of library. Breaking changes. 

From now it is based on json-bigint (#11 ).

Internally, our json representation (Aeson) relies on BigNumber from bignumber.js, to represent numbers.
This allows us deserialize/serialize JSONs holding arbitrary size numbers, including numbers with floating point.

### CODE REVIEW

**Review by klntsky — CHANGES_REQUESTED** [review]

**Comment by uhbif19** [comment]
https://github.com/mlabs-haskell/purescript-aeson/issues/12 will be probably closed too by this PR.

File: src/Aeson.purs [hunk 1]
```diff
@@ -451,85 +446,82 @@ toArray = toAesonType caseAesonArray
...
 
-decodeNumber
-  :: forall a. (String -> Maybe a) -> Aeson -> Either JsonDecodeError a
-decodeNumber parse aeson@(Aeson { numberIndex }) = do
-  -- Numbers are replaced by their index in the array.
-  ix <- decodeAesonViaJson aeson
-  numberStr <- note MissingValue (Seq.index ix numberIndex)
-  note (TypeMismatch $ "Couldn't parse to integral: " <> numberStr)
-    (parse numberStr)
+-- decodeNumber ∷ ∀ a. (String → Maybe a) -> String -> Aeson -> Either JsonDecodeError a
```
Comment (uhbif19) [inline src/Aeson.purs:458]: Commented code.

File: src/Aeson.purs [hunk 2]
```diff
@@ -1,213 +1,178 @@
...
--- |
--- | Then, in decoders for `Int` and `BigInt` we access that array to get the
--- | values back
--- |
--- | Known limitations: does not support Record decoding (no GDecodeJson-like
--- | machinery). But it is possible to decode records manually, because
--- | `getField` is implemented.
-module Aeson
-  ( (.:)
+-- | Uses json-bigint library to parse JSON.
```
Comment (uhbif19) [inline src/Aeson.purs:1]:
```suggestion
-- | Uses `json-bigint` library to parse JSON.
```

**Review by uhbif19 — CHANGES_REQUESTED** [review]

**Comment by uhbif19** [comment]
@klntsky @kirill-havryliuk Resolved already covered comments.

File: src/Aeson.purs [hunk 3]
```diff
@@ -153,15 +153,14 @@ foreign import stringifyAeson :: Aeson -> String
...
-toStringifiedNumbersJson = fix \_ ->
-  caseAeson
-    { caseNull: const Argonaut.jsonNull
-    , caseBoolean: encodeBoolean
-    , caseBigNumber: encodeString <<< BigNumber.toFixed
-    , caseString: encodeString
-    , caseArray: map toStringifiedNumbersJson >>> Argonaut.fromArray
-    , caseObject: map toStringifiedNumbersJson >>> Argonaut.fromObject
-    }
+toStringifiedNumbersJson = defer $ \_ -> caseAeson
```
Comment (klntsky) [inline src/Aeson.purs:156]:
```suggestion
toStringifiedNumbersJson = defer \_ -> caseAeson
```

**Review by klntsky — CHANGES_REQUESTED** [review]
Same everywhere where applicable

File: src/Aeson.purs (hunk 3 above)
Comment (klntsky) [reply src/Aeson.purs:156]: https://github.com/Plutonomicon/cardano-transaction-lib/issues/56#issue-1116215499

**Comment by uhbif19** [comment]
@klntsky @kirill-havryliuk Looks like all issues resolved, no? Merge it?

**Comment by kirill-havryliuk** [comment]
> @klntsky @kirill-havryliuk Looks like all issues resolved, no? Merge it?

@uhbif19 I would like to refactor CTL so it is based on this version first, so if we miss something we still can add it here, and not introduce new PR.

File: src/Aeson.purs [hunk 4]
```diff
@@ -406,8 +415,9 @@ caseAesonBigInt def f =
         else Nothing
     )
 
--- | The reson we return `Finite BigNumber` instead plain `BigNumber`
--- | is to simplify things like `caseAesonBigNumber _ fromFiniteBigNumber`
+-- | The reson we return `Finite BigNumber` instead of a plain `BigNumber`,
```
Comment (uhbif19) [inline src/Aeson.purs:418]:
```suggestion
-- | The reason we return `Finite BigNumber` instead of a plain `BigNumber`,
```

File: src/Aeson.js [hunk 5]
```diff
@@ -46,33 +46,45 @@ exports._caseAeson = _caseAeson
...
-        return tmp
-    })
-
-exports.stringifyAeson = json => JSONbig.stringify(json)
+// This one is not stack-safe and can
+// throw "Maximum call stack size exceeded" for big json's in practice
+// thus, commented until we found better way to perform this.
+
+const traverseFormattingBigNumber = identity
+// _caseAeson
```
Comment (uhbif19) [inline src/Aeson.js:74]: Do not understand what happened.

File: test/Gen.purs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      go =<< chooseInt 1 size
+      where
+      go _size = do
+        i <- intStr 0 top
+        let lenI = S.length i
+        if lenI >= _size then pure $ S.take _size i
+        else map (i <> _) $ go (_size - lenI)
+
+  negSign <- arbitrary
+  -- forceSign <- arbitrary
```
Comment (klntsky) [inline test/Gen.purs:96]: Leftover?

File: test/Gen.purs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  expPart <- largeIntStr expDigitsUpTo
+
+  let
+    number = when negSign "-" -- sign
+
+      <> intPart
+      <> when withFractional ("." <> fracPart)
+      <> when withExp ("e" <> expSign <> expPart)
+      where
+      -- sign = if negSign then "-" else when forceSign "+"
```
Comment (klntsky) [inline test/Gen.purs:115]: Leftover?

**Review by klntsky — APPROVED** [review]

**Pull request merged**